### ISSUE TITLE
[Question] Split all call chain in Java code

### ISSUE THREAD

**Issue by backcover7** [INRIA/spoon#5008]
Hi
I just started using spoon and I am not sure if spoon is able to do something in my requirements.

This is code snippet and it imported 3pp maven repo `org.codehaus.groovy.groovy 1.8.6`
```java
import org.codehaus.groovy.jsr223.GroovyScriptEngineFactory;
import org.codehaus.groovy.jsr223.GroovyScriptEngineImpl;

import javax.script.ScriptEngineManager;
import javax.script.ScriptException;

public class callchain {
    public static void main(String[] args) throws ScriptException {
        new GroovyScriptEngineFactory().getScriptEngine().eval("javacode");
        new GroovyScriptEngineImpl().getFactory().getScriptEngine().eval("javacode");
        new ScriptEngineManager().getEngineByName("xx").eval("javacode");
    }
}
```

As you can see there are three expressions in the main function and they are all calling eval() function which is defined in `javax.script.ScriptEngine` class in JDK at last. But you cannot find the `javax.script.ScriptEngine` class directly in the Java code because the instantiated object of the class is implicitly generated in the call chain.

**The question is**,
Can I use spoon to get the classtype of every method's caller object in such cases?

**Comment by algomaster99** [comment]
Hi @backcover7 !

> Can I use spoon to get the classtype of every method's caller object in such cases?

Do you mean that you need to know in which class the method is declared? For example, `eval` is declared in `javax.script.ScriptEngine`. You can achieve this with Spoon like so:
```java
Launcher launcher = new Launcher();
launcher.addInputResource(<path to file>);
CtModel model = launcher.buildModel();

List<CtExecutableReference> executableReferences = model.getElements(new TypeFilter<>(CtExecutableReference.class))
```
Then you can call `.getDeclaringType()` on each element in `executableReferences`. If you get `null` for your methods, you can try passing the classpath to the project before you build the model - `launcher.getEnvironment().setSourceClasspath(String[])`.

**Comment by backcover7** [comment]
@algomaster99 It works for me! Thank you!

**Comment by backcover7** [comment]
@algomaster99 I got another question, it seems I cannot find exact line number of CtExecuteReference. The getPosition() always return unknown file. How can I get the line number of the element?

**Comment by I-Al-Istannen** [comment]
That's a bit of a tricky thing. The executable ref has no source position (which is probably a missing feature), but the parent of it (the invocation) has one.

Currently spoon seems to do the following:
```java
foo.bar(0, 20)
^^^            target     position
        ^  ^^  literal    positions
^^^^^^^^^^^^^^ invocation position
```

I think the following would be slightly nicer:
```java
foo.bar(0, 20)
^^^            target         position
        ^  ^^  literal        positions
    ^^^        executable ref position
^^^^^^^^^^^^^^ invocation     position
```

I might create a PR changing that.

----

You need to be aware that you can't get "the line of an element", as many elements span multiple lines:
```java
foo()
  .bar()
  .baz()
```
here the `.baz()` invocation spans three lines: Its target (`foo().bar()`) spans two and the executable ref + arguments another one.

**Comment by backcover7** [comment]
@algomaster99 After I get the ctExecutableReference, I want to request its actual declaration type. The method `ctExecutableReference.getDeclaringType()` can do that. But I cannot get an override type.

For example. for the following java code
```java
public class test {
    public void x(JsonParser p, Class<T> valueType throws NamingException {
        new com.fasterxml.jackson.dataformat.cbor.databind.CBORMapper().readValue(p, valueType);
    }
}
```

The hierarchy is like the following. The CBORMapper is subclass of ObjectMapper and ObjectMapper is subclass of ObjectCodec.
```
Object (java.lang)
  TreeCodec (com.fasterxml.jackson.core)
    ObjectCodec (com.fasterxml.jackson.core)
      ObjectMapper (com.fasterxml.jackson.databind)
        YAMLMapper (com.fasterxml.jackson.dataformat.yaml)
        CBORMapper (com.fasterxml.jackson.dataformat.cbor.databind)
        JsonMapper (com.fasterxml.jackson.databind.json)
        JsonMapper (com.fasterxml.jackson.databind.json)
```
The `getDeclaringType` will get `ObjectCodec` **BUT** actually in the `ObjectMapper` the `readValue` method has been override. so I do not want to get `ObjectCodec` if the method has been overridden by the one of the superclass below the `ObjectCodec`. How can I do that?

**Comment by I-Al-Istannen** [comment]
Hey, the following works:

<details>

<summary>A smaller example with full source</summary>

```java
	@Test
	void foo() {
		Launcher launcher = new Launcher();
		launcher.addInputResource(new VirtualFile(
			"import java.util.*;\n" +
				"\n" +
				"public class Foo<T> extends ArrayList<T> {\n" +
				"  public void entrypoint() {\n" +
				"    size();\n" +
				"  }\n" +
				"\n" +
				"  @Override\n" +
				"  public int size() {\n" +
				"    return 42;\n" +
				"  }\n" +
				"}",
			"Foo.java"
		));
		CtType<?> foo = launcher.buildModel().getAllTypes().iterator().next();
		CtInvocation<?> invocation = foo.getMethodsByName("entrypoint").get(0).getBody().getStatement(0);
		CtMethod<?> executable = (CtMethod<?>) invocation.getExecutable().getExecutableDeclaration();
		System.out.println(executable);
		System.out.println(executable.getDeclaringType().getQualifiedName());
	}

```
```
@java.lang.Override
public int size() {
    return 42;
}
Foo
```

</details>

<details>

<summary>Your example</summary>

```java
  public static void foo() {
    Launcher launcher = new Launcher();
    launcher.getEnvironment().setComplianceLevel(17);
    launcher.getEnvironment().setNoClasspath(false);
    launcher.addInputResource(new VirtualFile(
        """
            import com.fasterxml.jackson.core.JsonParser;
            public class Test {
                public void x(JsonParser p, Class<?> valueType) throws Exception {
                    new com.fasterxml.jackson.dataformat.cbor.databind.CBORMapper().readValue(p, valueType);
                }
            }""",
        "Test.java"
    ));
    CtType<?> test = launcher.buildModel().getAllTypes().iterator().next();
    CtMethod<?> x = test.getMethodsByName("x").get(0);
    CtInvocation<?> invocation = x.getBody().getStatement(0);
    CtMethod<?> method = (CtMethod<?>) invocation.getExecutable().getExecutableDeclaration();
    System.out.println(method);
    System.out.println(method.getDeclaringType().getQualifiedName());
  }
```

```
public <T> T readValue(com.fasterxml.jackson.core.JsonParser arg0, java.lang.Class<T> arg1) throws java.io.IOException, com.fasterxml.jackson.databind.DatabindException, com.fasterxml.jackson.core.exc.StreamReadException {
}
com.fasterxml.jackson.databind.ObjectMapper
```

Using
```java
invocation.getExecutable().getDeclaringType()
```
should work as well :)

</details>

As far as I can see those are all correct, `CBORMapper` does not override `ObjectMapper#readValue`. Are you sure that you have the relevant classes in your classpath or sourcepath?

**Comment by backcover7** [comment]
@I-Al-Istannen I must have confused you, `CBORMapper` did not override the `ObjedtMapper#readValue`. I just wanted to get `ObjectMapper#readValue` instead of  `ObjectCodec#readValue`.

And is there any way to convert `CtExecuteReference` to `CtInvocation`? Or could you suggest me a way to traverse all executable references and get their actual declaration or override declaration?

**Comment by backcover7** [comment]
> @I-Al-Istannen I must have confused you, `CBORMapper` did not override the `ObjedtMapper#readValue`. I just wanted to get `ObjectMapper#readValue` instead of `ObjectCodec#readValue`.
> 
> But I cannot get the `invocation` through a statement like `Object a = cborMapper.readValue(p, valueType);` because it is a `CtLocalVariable` object and it is complex to parse some other possible cases like `CtAssignment` to get **every** `CtInvocation` in it. How can I do that?
> 
> And is there any way to convert `CtExecuteReference` to `CtInvocation`? Or could you suggest to me a way to traverse all executable references including invocation and constructor and get their actual declaration or override declaration?

```java
List<CtConstructorCall<?>> constructorCallList = method.getElements(new TypeFilter<>(CtConstructorCall.class));
for (CtConstructorCall<?> constructorCall : constructorCallList) {
    CtConstructor<?> executableInvocation = (CtConstructor<?>) constructorCall.getExecutable().getExecutableDeclaration();
    System.out.println(executableInvocation.getDeclaringType().getQualifiedName());
}

List<CtInvocation<?>> ctInvocationList = method.getElements(new TypeFilter<>(CtInvocation.class));
for (CtInvocation<?> invocation : ctInvocationList) {
    CtMethod<?> executableInvocation = (CtMethod<?>) invocation.getExecutable().getExecutableDeclaration();
    System.out.println(executableInvocation.getDeclaringType().getQualifiedName());
}
```

Found a way myself~

**Comment by I-Al-Istannen** [comment]
>  I must have confused you, `CBORMapper` did not override the `ObjedtMapper#readValue`. I just wanted to get `ObjectMapper#readValue` instead of  `ObjectCodec#readValue`.

But that is what already happens in my examples? It returns `ObjectMapper#readValue` and that should be correct as `CBORMapper` does not specialize further.

> Found a way myself~

Nice :) You can also filter for `CtAbstractInvocation` which encompasses invocations, constructor calls and anonymous class creations.

**Comment by I-Al-Istannen** [comment]
After #5011 executable references now have a source position. I think this issue can therefore be closed, feel free to comment/reopen if you still have any questions :)

**Comment by I-Al-Istannen** [comment]
Spoon returns you the method called on the *static* type (as that one is known statically) and not the *dynamic* type (which will be there at runtime). 

The simple reason is that **this problem is undecidable**. Your analysis will *never* work correctly in all cases, so you will need to carefully think about what cases you want to support. An easy example to see why you won't necessarily be correct is:
```java
Context ctx;
if (System.in.read() == 32) {
  ctx = new Foo();
} else {
  ctx = new Bar();
}
```
or, if you do not want to depend on a human:
```java
Context ctx;
if ("hello".isBlank()) { // make this as complex as needed so your analysis isn't sure anymore
  ctx = new Foo();
} else {
  ctx = new Bar();
}
```
To solve this you would need to figure out what code is dead and that the possible values are at this point — and that involves some unpleasant interactions with the halting problem.

----

In your *specific* pattern, you should be able to:
1. get the `target` of the invocation, which is a local variable reference for `ctx`
2. get its local variable declaration using the corresponding method
3. use `getAssignment` to get the assignment
4. `getType` the type of the assignment expression
5. lookup the method with the corresponding signature in that type. You already know the name and parameter types from your invocation, so `CtType#getMethod(String, CtTypeReference<?>...)` should be enough.

**Comment by backcover7** [comment]
@I-Al-Istannen 

I got another question when parsing invocation. When traversing the source code like `Runtime.getRuntime().exec()`, spoon returns an invocation. And I tried to acquire the source code block using `invocation.toString()` but it returns `java.lang.Runtime.getRuntime().exec()` since `Runtime` is a class under `java.lang` package.

However it is not precise, I only want to get exact original source code. How can I get the unmodified code from the invocation object?

**Comment by I-Al-Istannen** [comment]
If you are indexing an actual file, `invocation.getOriginalSourceFragment().getSourceCode()` should work. It won't for virtual files or zips (I think) as it reads in the file contents and then uses the start and end offset stored in the element's position to substring the original source code.

It logically also won't work for implicit elements (like `super()` invocations), as those do not exist in the file.

What do you need the original code for? Maybe there is a nicer way to do what you want that does not involve reproducing the original source code.

**Comment by backcover7** [comment]
@I-Al-Istannen I just need to record some required source code as log.
When I parsed such invocation
```java
//        String inte = s.intermediate("m");
        s.rce(user);
```
It will throw an exception when calling getOriginalSourceFragment().
And the result of `invocation.toString()` contains the comment like `// String inte = s.intermediate("m");\ns.rce(user)`

I know how to remove the comment by using `filterChildren()` and `removeComment()`
But I assume the comment is not part of the invocation. Is there any simple way to get the source code without comment?

**Comment by I-Al-Istannen** [comment]
It would be interesting to know why it fails for you (do you add a file or zip as input resource)? What does `invocation.getPosition()` return?

Comments are metadata, not part of the AST. You can stop Spoon from attaching them by setting `setCommentEnabled` to `false` in `launcher.getEnvironment()` before building the model.

**Comment by backcover7** [comment]
the `setCommentEnabled=false` works
